ibis/expr/types/relations.py

```diff
@@ -1755,10 +1756,16 @@ def union(self, table: Table, *rest: Table, distinct: bool = False) -> Table:
         │     3 │
         └───────┘
         """
-        node = ops.Union(self, table, distinct=distinct)
-        for table in rest:
-            node = ops.Union(node, table, distinct=distinct)
-        return node.to_expr().select(self.columns)
+        queue = deque()
+        queue.append(self)
+        queue.append(table)
+        queue.extend(rest)
+        while len(queue) > 1:
+            node1 = queue.popleft()
+            node2 = queue.popleft()
+            union_node = ops.Union(node1, node2, distinct=distinct)
+            queue.append(union_node)
+        return queue[0].to_expr().select(self.columns)
```

Comment:
HI @kszucs Thanks for your feedback.

It is still pairwise union, there is a slight difference to the original implementation, something like breadth first pairwise, it could reduce the depth in sql query. 

Here is the `expr` and `sql` in this new implementation.

```
 r0 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      5

r1 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0     -1

r2 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      0

r3 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      1

r4 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      2

r5 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      3

r6 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      4

r7 := Union[r1, r2, distinct=False]

r8 := Union[r3, r4, distinct=False]

r9 := Union[r5, r6, distinct=False]

r10 := Union[r0, r7, distinct=False]

r11 := Union[r8, r9, distinct=False]

r12 := Union[r10, r11, distinct=False]

Project[r12]
  col_0: r12.col_0
  ```
  
  ```sql
   SELECT
  *
FROM (
  SELECT
    *
  FROM (
    SELECT
      *
    FROM "ibis_pandas_memtable_3dqxlaaxrjbrhn6sezkq3jfey4" AS "t0"
    UNION ALL
    SELECT
      *
    FROM (
      SELECT
        *
      FROM "ibis_pandas_memtable_gjf7ximrfnat7ocq7u3snufdwy" AS "t1"
      UNION ALL
      SELECT
        *
      FROM "ibis_pandas_memtable_khxo6jdnpnbc3lz4kdpkoll6ny" AS "t2"
    ) AS "t7"
  ) AS "t10"
  UNION ALL
  SELECT
    *
  FROM (
    SELECT
      *
    FROM (
      SELECT
        *
      FROM "ibis_pandas_memtable_4wg3k2vsvzf4tophotkj6n6ope" AS "t3"
      UNION ALL
      SELECT
        *
      FROM "ibis_pandas_memtable_zpp6dkndfnhjla473tjklyp3ku" AS "t4"
    ) AS "t8"
    UNION ALL
    SELECT
      *
    FROM (
      SELECT
        *
      FROM "ibis_pandas_memtable_y5w7wpkhkbgr3agcuc3ftnujai" AS "t5"
      UNION ALL
      SELECT
        *
      FROM "ibis_pandas_memtable_ebrddatk6bfgti4x7gwrkgi3gm" AS "t6"
    ) AS "t9"
  ) AS "t11"
) AS "t12"
```

Here is the expr and sql in the original one:

```
 r0 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      5

r1 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      4

r2 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      3

r3 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      2

r4 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      1

r5 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0     -1

r6 := InMemoryTable
  data:
    PandasDataFrameProxy:
         col_0
      0      0

r7 := Union[r5, r6, distinct=False]

r8 := Union[r7, r4, distinct=False]

r9 := Union[r8, r3, distinct=False]

r10 := Union[r9, r2, distinct=False]

r11 := Union[r10, r1, distinct=False]

r12 := Union[r11, r0, distinct=False]
```
original sql
```
 SELECT
  *
FROM (
  SELECT
    *
  FROM (
    SELECT
      *
    FROM (
      SELECT
        *
      FROM (
        SELECT
          *
        FROM (
          SELECT
            *
          FROM (
            SELECT
              *
            FROM "ibis_pandas_memtable_n4qzohbx3nfjdi6crezpa3zkfu" AS "t5"
            UNION ALL
            SELECT
              *
            FROM "ibis_pandas_memtable_3dhomfxvdfb2tezjssmbdkvvam" AS "t6"
          ) AS "t7"
          UNION ALL
          SELECT
            *
          FROM "ibis_pandas_memtable_mqknkr6dr5gv7hpcrri34hxydq" AS "t4"
        ) AS "t8"
        UNION ALL
        SELECT
          *
        FROM "ibis_pandas_memtable_v5wru54hhfabxd7khavq27zruq" AS "t3"
      ) AS "t9"
      UNION ALL
      SELECT
        *
      FROM "ibis_pandas_memtable_m4hgmj33nreopoxvxwhkycicfy" AS "t2"
    ) AS "t10"
    UNION ALL
    SELECT
      *
    FROM "ibis_pandas_memtable_lilxpxko6jh4lplqw2kp6jny3y" AS "t1"
  ) AS "t11"
  UNION ALL
  SELECT
    *
  FROM "ibis_pandas_memtable_mxiz4rs5z5dohdjxror5hm6p5a" AS "t0"
) AS "t12"
```

Reply:
This seems like a worthwhile improvement.